### PR TITLE
Add documentation for running a reindex

### PR DESCRIPTION
**Pull request by kenoir**
The process of running a reindex requires many steps and is not documented. This change adds some documentation that would allow a developer unfamiliar with the process to make a decent attempt to understand and implement a reindex.

[View the rendered version](https://github.com/wellcomecollection/catalogue-pipeline/blob/add-reindex-docs/REINDEXING.md).

This fulfills one of the actions of the post-mortem retrospective here: https://github.com/wellcomecollection/docs/blob/main/incident_retros/2021-06-28_slow-search.md

### CODE REVIEW

File: REINDEXING.md [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+
+A reindex operation runs the source data from the [adapters](docs/adapters/README.md) through the pipeline causing it to be re-transformed / matched & merged as appropriate.
+
+## How to run a reindex
+
+Running a reindex will require:
+
+- Preparing and deploying a release
+- Terraforming a new pipeline using that release
+- Running the reindex script to feed the new pipeline
```
Comment (alexwlchan) [inline REINDEXING.md:11]: I would maybe break this down into a checklist.

So up here you have a checklist that somebody who’s done reindexes can refer to, and further down you have more detailed explanations of what's going on.

File: REINDEXING.md [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+
+### Update the catalogue-pipeline environments
+
+The catalogue-pipeline is deployed using [weco-deploy](https://github.com/wellcomecollection/weco-deploy) but does not deploy in the usual way as there is often one release per pipeline.
```
Comment (alexwlchan) [inline REINDEXING.md:15]:
```suggestion
The catalogue-pipeline is deployed using [weco-deploy](https://github.com/wellcomecollection/weco-deploy) but does not deploy in the usual way as there is often one environment per pipeline. 
```

File: REINDEXING.md [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+```
+catalogue_pipeline:
+  environments:
+    - id: "2021-06-27"
+      name: "2021-06-27"
+    - id: "2021-07-06"
+      name: "2021-07-06"
+    - id: "MY-NEW-PIPELINE"
+      name: "MY-NEW-PIPELINE"
+```
```
Comment (alexwlchan) [inline REINDEXING.md:30]: Suggestion:

<pre>
```diff
 catalogue_pipeline:
   environments:
     - id: "2021-06-27"
       name: "2021-06-27"
     - id: "2021-07-06"
       name: "2021-07-06"
+    - id: "YYYY-MM-DD"
+      name: "YYYY-MM-DD"
```
</pre>

which renders as:

```diff
 catalogue_pipeline:
   environments:
     - id: "2021-06-27"
       name: "2021-06-27"
     - id: "2021-07-06"
       name: "2021-07-06"
+    - id: "YYYY-MM-DD"
+      name: "YYYY-MM-DD"
```

to be super clear about what somebody should be adding.

File: REINDEXING.md [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+relation_embedder      021df38      -            -
+router                 021df38      -            -
+transformer_calm       021df38      -            -
+transformer_mets       021df38      -            -
+transformer_miro       021df38      -            -
+transformer_sierra     021df38      -            -
+transformer_tei        f9f910d      -            -
+
+Created release eeb9482f-0a6a-4211-8de9-27cee7567134
+```
```
Comment (alexwlchan) [inline REINDEXING.md:66]: I would trim the output just to show the important parts, and the command – the exact output isn't particularly relevant to running a reindex, and will get out-of-date quickly.

File: REINDEXING.md [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+
+  pipeline_date = "MY-NEW-PIPELINE"
+  release_label = "MY-NEW-PIPELINE"
+
+  is_reindexing = true
+
+  # Boilerplate that shouldn't change between pipelines.
+  # ...
+}
+```
```
Comment (alexwlchan) [inline REINDEXING.md:181]: At no point does this instruction explain that you need to plan/apply the pipeline – which is probably worth explaining explicitly (especially as the pipeline Terraform is a bit unusual).

File: REINDEXING.md [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+
+### Running the reindex script
```
Comment (alexwlchan) [inline REINDEXING.md:183]: Can we pick a better name for this operation? What does run-in the reindex script do?

File: REINDEXING.md [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+
+```
+> python3 ./start_reindex.py
+
+Which source do you want to reindex? (all, miro, sierra, mets, calm): all
+Which pipeline are you sending this to? (catalogue, catalogue_miro_updates, reporting): catalogue
+Every record (complete), just a few (partial), or specific records (specific)? (complete, partial, specific): partial
+```
+
+You may wish to run a partial-reindex to verify that the process will work as expected before embarking on a complete reindex.
```
Comment (alexwlchan) [inline REINDEXING.md:197]: What does a partial reindex do?

File: REINDEXING.md [hunk 8]
```diff
@@ -0,0 +1,261 @@
...
+
+You can monitor a reindex in progress using Grafana at [https://monitoring.wellcomecollection.org/](https://monitoring.wellcomecollection.org/), or by looking at CloudWatch metrics in the `platform` AWS account.
```
Comment (alexwlchan) [inline REINDEXING.md:199]: We also have the `get_reindex_status.py` script: https://github.com/wellcomecollection/catalogue-pipeline/blob/main/reindexer/get_reindex_status.py which reports:

- the total number of source records
- how many works have reached each stage of the reindex

File: REINDEXING.md [hunk 9]
```diff
@@ -0,0 +1,261 @@
...
+
+Non-empty DLQs will be reported in the Wellcome #wc-platform-alerts Slack channel.
+
+A reindex should take less than a working day to complete. 
+
+### Connecting the Catalogue API
+
+When we have a new index in the Elasticsearch cluster created for the pipeline when we terraformed it, we will want to make that index available to the Catalogue API cluster.
```
Comment (alexwlchan) [inline REINDEXING.md:207]: I know what you mean, but cramming it all into one sentence is a bit unwieldy.

Also, be clear about when we want to do this:

```suggestion
Every pipeline has a dedicated Elasticsearch cluster, and this cluster contains indexes for different stages of the reindex.
The two indexes of interest are `works-indexed-{PIPELINE_DATE}` and `images-indexed-{PIPELINE_DATE}` – these contain documents suitably for the API to use.

When these indexes are fully populate, we want to replicate them into the `catalogue-api` cluster.
```

And possibly a diagram would help?

File: REINDEXING.md [hunk 10]
```diff
@@ -0,0 +1,261 @@
...
+    ElasticConfig(
+      worksIndex = Index(s"works-indexed-$indexDate"),
+      imagesIndex = Index(s"images-indexed-$indexDate")
+    )
+}
+```
+
+You will want to deploy this change through the API stage environment and allow CI to perform the usual API checks. 
+
+Be sure to check the diff_tool output in CI before deploying to production!
```
Comment (alexwlchan) [inline REINDEXING.md:259]: How do I do this?

File: REINDEXING.md [hunk 11]
```diff
@@ -0,0 +1,261 @@
...
+
+  def apply(): ElasticConfig =
+    ElasticConfig(
+      worksIndex = Index(s"works-indexed-$indexDate"),
+      imagesIndex = Index(s"images-indexed-$indexDate")
+    )
+}
+```
+
+You will want to deploy this change through the API stage environment and allow CI to perform the usual API checks.
```
Comment (alexwlchan) [inline REINDEXING.md:257]: How do I do this?

File: REINDEXING.md (hunk 2 above)
Comment (kenoir) [reply REINDEXING.md:15]: I'm going to omit the bit "but does not deploy in the usual way" is it begs explanation which would be unnecessary detail here. I'm going to state only "Each environment represents a new pipeline."

File: REINDEXING.md (hunk 10 above)
Comment (kenoir) [reply REINDEXING.md]: I feel like there should probably be another guide for deploying the catalogue API to refer to, but there isn't - so i'll add some sparse detail here.

File: REINDEXING.md (hunk 11 above)
Comment (kenoir) [reply REINDEXING.md]: Same as before for catalogue-api docs about deployment.

**Review by jamieparkinson — APPROVED** [review]
This is very good :tada:

File: REINDEXING.md [hunk 12]
```diff
@@ -0,0 +1,279 @@
...
+
+Give your new pipeline environment id and name using the date pattern `YYYY-MM-DD`.
+
+You will need to use the environment id when terraforming a new pipeline.
+
+### Prepare a release
+
+Prepare a release following the example below. 
+
+You may want to choose a specific git ref as a label and _not_ `latest`.
```
Comment (alicefuzier) [inline REINDEXING.md:43]: ha I did not know you could use the git ref

**Review by alicefuzier — COMMENTED** [review]
👍  looks good and really useful, thanks

**Pull request merged**